client/control/components/RecipeForm.js

```diff
@@ -1,197 +1,206 @@
-              </Link>
-            }
-          </div>
-          <div className="fluid-2 float-right">
-            <button className="button" type="submit" disabled={submitting} >Submit</button>
-          </div>
+        <ControlField label="Action" name="action" component="select">
+          <option value="">Choose an action...</option>
+          <option value="console-log">Log to Console</option>
+          <option value="show-heartbeat">Heartbeat Prompt</option>
+        </ControlField>
+        {ArgumentsFields && <ArgumentsFields />}
+        <div className="form-actions">
+          {recipeId &&
+            <Link className="button delete" to={`/control/recipe/${recipeId}/delete/`}>
+              Delete
+            </Link>
+          }
+          <button className="button submit" type="submit" disabled={submitting}>Submit</button>
         </div>
-
       </form>
     );
   }
 }
 
-export default composeRecipeContainer(reduxForm({
+/**
+ * Redux-Form config for the RecipeForm.
+ */
+export const formConfig = {
   form: 'recipe',
-}, (state, props) => {
-  const fields = ['name', 'filter_expression', 'enabled', 'action'];
-  const selectedRecipeRevision = (props.location.state)
-    ? props.location.state.selectedRevision
-    : null;
-
-  const formatErrors = payload => {
-    let errors = payload;
-
-    /* If our payload is an object, process each error in the object
-       Otherwise, it is a string and will be returned immediately */
-    if (_.isObject(payload)) {
-      const invalidFields = Object.keys(payload);
-      if (invalidFields.length > 0) {
-        /* If our error keys are integers, it means they correspond
-           to an array field and we want to present errors as an array
-           e.g. { surveys: {0: {title: 'err'}}, {2: {weight: 'err'}} }
-           =>   { surveys: [{title: 'err'}, null, {weight: 'err'}] } */
-        errors = isNaN(invalidFields[0]) ? {} : [];
-
-        invalidFields.forEach(fieldName => {
-          errors[fieldName] = formatErrors(payload[fieldName]);
-        });
+  asyncBlurFields: ['filter_expression'],
+
+  async asyncValidate(values) {
+    const errors = {};
+
+    // Validate that filter expression is valid JEXL
+    if (!values.filter_expression) {
+      errors.filter_expression = 'Filter expression cannot be empty.';
+    } else {
+      const jexlEnvironment = new JexlEnvironment({});
+      try {
+        await jexlEnvironment.eval(values.filter_expression);
+      } catch (err) {
+        errors.filter_expression = err.toString();
       }
     }
 
-    return errors;
-  };
-
-  const onSubmitFail = errors => {
-    const { dispatch } = props;
-    const actionFormErrors = errors.arguments;
+    // Throw if we found any errors.
+    if (Object.keys(errors).length > 0) {
+      throw errors;
+    }
+  },
+
+  onSubmit(values, dispatch, { recipeId, updateRecipe, addRecipe }) {
+    // Filter out unwanted keys for submission.
+    const recipe = _.pick(values, [
```

Comment:
How do you feel about changing `import _ from 'underscore'` to `import { pick } from 'underscore'`, and just going with `... = pick(values, ...`?

If we ever switch to Webpack 2, this would drop the bundle size after tree shaking